### ISSUE TITLE
Compile each benchmark function into its own file

### ISSUE THREAD

**Issue by yamiteru** [yamiteru/isitfast#44]
## Problem

Right now I compile the whole file and then import the function. 

But calling the benchmark function leaves an additional performance overhead which is why up to this point we've been collecting "offsets" which is basically just a benchmark of an empty a/sync function to get the cost of calling an empty function.

This has proven difficult and unpredictable. I've looked more deeply into how other similar libraries handle this issue. Most of them don't but those that do almost universally unwrap/inline the function.

But all of the other libraries work like `isitfast` has worked before turning it into CLI-first app. In those cases it's relatively easy to extract the function, surround it with measurement functions and call it a day.

However in our case it's a bit more trickier because we can combine implementation and benchmark in a single file like this:

```ts
const mergeObjects = (...objects: ObjectAny[]) => { /* .. */ };

export const $$benchmark = () => { /* .. */ };
```

Here we have to preserve the scope. This in itself is still kinda easy but what happens when the scope looks like this?

```ts
const mergeObjects = (...objects: ObjectAny[]) => { /* .. */ };

const tmpBenchmarkData = { /* .. */ };

const helpFunction = () => { /* .. */ };

export const $benchmarkOne = () => { /* .. */ };

export const $benchmarkTwo = () => { /* .. */ };
```

First we identify that there are two benchmarks in the file. We want to have as accurate results as possible and all of those un-needed constants and function might introduce statistical noise since they will also consume some cpu time and memory.

So ideally what should happen is that we should get these two files (`helpFunction` is used only in `$benchmarkTwo`):

```ts
// benchmarkOne.js

const mergeObjects = (...objects: ObjectAny[]) => { /* .. */ };

const tmpBenchmarkData = { /* .. */ };

export const $benchmarkOne = () => { 
  const start = process.hrtime.bigint();
  
  /* .. */ 

  return Number(process.hrtime.bigint() - start);
};
```

```ts
// benchmarkTwo.js

const mergeObjects = (...objects: ObjectAny[]) => { /* .. */ };

const tmpBenchmarkData = { /* .. */ };

const helpFunction = () => { /* .. */ };

export const $benchmarkOne = () => { 
  const start = process.hrtime.bigint();
  
  /* .. */ 

  return Number(process.hrtime.bigint() - start);
};
```

This is a non-trivial compilation that requires to parse the code, get dependency graph, add measurement code and write the result into a file. Also since we're measuring cpu and ram separately then each benchmark should produce two files with different measurement code.

## Solution

I have to get AST by using `parseFile` from `@swc/core`, iterate over it, create a new AST tree with only the nodes I need and then use `print` to get the JS code and save it into a file.

File names should look like this `${uuid}_${mode}`.

**Comment by yamiteru** [comment]
So I have the basic functionality working but I'm gonna make a few changes to the solution.

First of all I'm not gonna be compiling away the code that I don't need since it'd require much more time playing around with AST which I don't have right now.

Secondly even if I did I would still be getting memory noise from for example the setup/shared state. So I'm gonna bring back offsets but only for the memory. I'm still not 100% sure how/when exactly I'm gonna collect them and I'm too sleepy to think right now (it's 2:45 am).

Also for the sake of simplicity I'm gonna support only named function exports for now.

Lastly I'm gonna try adding best and worst case benchmarks by either enforcing optimizations or disabling them altogether. This means that I'll generate 4 files per benchmark (cpu-opt, cpu-deopt, ram-opt, ram-deopt).

I believe that the more data we have the more accurate stats we can have and the more meaningful results we can get.

**Comment by yamiteru** [comment]
I've figured out that even when I inline measurement code right into the benchmark functions it's still not gonna give me zero nanoseconds when I do `process.hrtime.bigint()`. So I'll probably have to fully reintroduce offsets. But I'm gonna do it in a separate issue and somewhere at the end of the development towards v1.0.

Secondly I've implemented the opt/normal/deopt so I have 6 files per benchmark.

Also I'm gonna get rid of unnecessary benchmark code later in a different issue since it's really gonna take a lot of time to do it right.

The algorithm works like this right now:

1. Collect force-optimized benchmarks (250 iterations)
2. Collect normal benchmarks (as many iterations as needed to reach less than or equal `force-optimized.median` for `Math.ceil(250 * 0.75)` iterations)
3. Collect force-deoptimized benchmarks (250 iterations)

The logic for stopping normal benchmark collection is very trivial as of now and might cause the benchmark to run infinitely so I'm gonna improve it in the future. 

We have a timeout to combat this issue but it's set to a static value of 10s but I'd like to set it dynamically based on how long the force-optimized version took.
